apis/duck/v1/knative_reference.go

```diff
@@ -86,3 +113,53 @@ func (kr *KReference) SetDefaults(ctx context.Context) {
 		kr.Namespace = apis.ParentMeta(ctx).Namespace
 	}
 }
+
+// ResolveGroup resolves the APIVersion of a KReference starting from the Group.
+// In order to execute this method, you need RBAC to read the CRD of the Resource referred in this KReference.
+// Note: This API is EXPERIMENTAL and might break anytime. For more details: https://github.com/knative/eventing/issues/5086
+func (kr *KReference) ResolveGroup(crdLister apiextensionsv1lister.CustomResourceDefinitionLister) error {
```

Comment:
I would not add this code to pkg nor kref. We intend to replace this lookup code with a lookup from ClusterDuckType. Hardcoding it here seems like a mistake. I would move this to a utility package to make it clear this is one method for resolving version from group.

This method has a side effect that invalidates the validation of the KRefrence in the validate method. That seems non-ideal as well. I think it should return a new KRef that has the alternante valid Kind+APIVersion set.

Reply:
> We intend to replace this lookup code with a lookup from ClusterDuckType. Hardcoding it here seems like a mistake

TBH I don't know what you're talking about, nor what ClusterDuckType is :smile:, can you be more specific about it?

> I would move this to a utility package to make it clear this is one method for resolving version from group.

Makes sense, which package do you propose? How about `knative.dev/pkg/krefutil`? Or maybe just `knative.dev/pkg/kref`? We could even more there some methods to convert KReference back and forth to other kube types

Reply:
Check the last commit, I moved to `knative.dev/pkg/kref`